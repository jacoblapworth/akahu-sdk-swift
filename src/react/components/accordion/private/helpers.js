import verge from 'verge';
import {calcSpaceRight, calcSpaceLeft} from '../../positioning/private/dom-helpers';

const getSpacesHorizontal = rect => ({
	spaceLeft: calcSpaceLeft(rect),
	spaceRight: calcSpaceRight(rect),
});

const hasEnoughSpaceHorizontal = ({ spaceLeft, spaceRight }) => {
	const MARGIN_POP = 8;
	const TOLERANCE = 1;
	const WIN_SCROLL_BAR = 17;

	return (
		spaceLeft > MARGIN_POP + TOLERANCE && spaceRight > MARGIN_POP + TOLERANCE + WIN_SCROLL_BAR
	);
};

export const shouldAccordionPop = rect => hasEnoughSpaceHorizontal(getSpacesHorizontal(rect));
export const getRectangle = rect => verge.rectangle(rect);
