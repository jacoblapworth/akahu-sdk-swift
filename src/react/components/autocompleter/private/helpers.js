
/**
 * Simple predicated used to determine if the ListBox's root DOM node can actually be focused.
 *
 * @public
 * @param {HTMLElement} node
 *
 * @returns {boolean}
 */
export const isVisible = node => node && window.getComputedStyle(node).visibility !== 'hidden';

/**
 * Attempts to run the passed predicate, if successful it will run the callback method and clear. If unsuccessful, it
 * will run several times over five seconds. If it's still unsuccessful after that it will try again several times over
 * a half second. If it's still not successful it will stop trying.
 *
 * @public
 * @param {Function} predicate
 * @param {Function} callback
 */
export const intervalRunner = (predicate, callback) => {
	if (predicate()) {
		callback();
	} else {
		let counter = 0;
		let interval = 0;
		const maxCount = 5;
		const delay = 100;
		const checker = function () {
			const check = predicate();
			if (check) {
				callback();
			}
			if (check|| counter > maxCount) {
				clearInterval(interval);
			}
			counter++;
		};

		interval = setInterval(checker, delay);
	}
};

export const dropdownSizeClasses = {
	small: 'ac-dropdown-small',
	medium: 'ac-dropdown-medium',
	large: 'ac-dropdown-large',
	xlarge: 'ac-dropdown-xlarge'
};
