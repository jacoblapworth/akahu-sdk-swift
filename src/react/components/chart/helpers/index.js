import getGroupPosition from './groupposition';
import createChartPadding from './chartpadding';

const alwaysPositive = value => Math.max(0, value) || 0;
const testIsCloseEnough = (next, previous, threshold = 2) => (
	next > previous - threshold && next < previous + threshold
);

export {getGroupPosition as default, createChartPadding, alwaysPositive, testIsCloseEnough};
