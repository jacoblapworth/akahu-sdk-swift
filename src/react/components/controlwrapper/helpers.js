import uuidv4 from 'uuid/v4';

/**
 * @public
 * Build the control, label, and message ids from a user-provided id, or from scratch
 * This should first be triggered by initialising the control component. The output is
 * then passed to getAriaAttributes.
 * @param {string} labelId - Optional, consumer-provided string to use as the label ID
 * @returns {{label: string, message: string, control: string}}
 */
export default function generateIds(labelId) {
  const label = labelId || uuidv4();
  return {
    label,
    control: `${label}-control`,
    message: `${label}-message`,
  };
}

/**
 * @public
 * Build the control, label, and message ids from a user-provided id, or from scratch
 * This should first be triggered by initialising the control component. The output is
 * then passed to getAriaAttributes.
 * @param {string} controlId - Optional, consumer-provided string to use as the control ID
 * @returns {{label: string, message: string, control: string}}
 */
export function generateIdsFromControlId(controlId) {
  const control = controlId || uuidv4();
  return {
    control,
    label: `${control}-label`,
    message: `${control}-message`,
  };
}
