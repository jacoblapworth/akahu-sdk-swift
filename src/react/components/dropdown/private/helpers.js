import breakpoints from '../../helpers/breakpoints';

/**
 * Simple predicated used to determine if the ListBox's root DOM node can actually be focused.
 *
 * @public
 * @param {HTMLElement} node
 *
 * @returns {boolean}
 */
export const isVisible = node => node && window.getComputedStyle(node).visibility !== 'hidden';


/**
 * Tests for height and width of the node to make sure it's rendered.
 *
 * @public
 * @param {HTMLElement} element
 *
 * @return {boolean}
 */
export const isRendered = element => element && element.offsetHeight > 0 && element.offsetWidth > 0;

/**
 * Attempts to run the passed predicate, if successful it will run the callback method and clear.
 * If unsuccessful, it will run several times over five seconds. If it's still unsuccessful after
 * that it will try again several times over a half second. If it's still not successful it will
 * stop trying.
 *
 * @public
 * @param {Function} predicate
 * @param {Function} callback
 */
export const intervalRunner = (predicate, callback) => {
	if (predicate()) {
		callback();
	} else {
		let counter = 0;
		let interval = 0;
		const maxCount = 5;
		const delay = 100;
		const checker = () => {
			const check = predicate();
			if (check) {
				callback();
			}
			if (check || counter > maxCount) {
				clearInterval(interval);
			}
			counter += 1;
		};

		interval = setInterval(checker, delay);
	}
};


/**
 * Detect if the item is below the bottom of the containing element that's scrollable
 * OR
 * Check it's not above the element that's scrollable.
 *
 * @public
 * @param {HTMLElement} itemDOM
 * @param {HTMLElement} scrollableDOM
 *
 * @returns {number}
 */
export function scrollTopPosition(itemDOM, scrollableDOM) {
	const elRect = itemDOM.getBoundingClientRect();
	const scrollableElRect = scrollableDOM.getBoundingClientRect();
	let scrollTopAmount = scrollableDOM.scrollTop;

	if (elRect.bottom > scrollableElRect.bottom) {
		scrollTopAmount += (elRect.bottom - scrollableElRect.bottom);
	} else if (elRect.top < scrollableElRect.top) {
		scrollTopAmount -= (scrollableElRect.top - elRect.top);
	}

	return scrollTopAmount;
}

export const checkIsNarrowViewport = () =>
	typeof window !== 'undefined'
	&& window.document.documentElement.clientWidth < breakpoints.small;

/**
 * Add the window event listeners that the DropDownToggled depends on.
 *
 * @export
 * @param {DropDownToggled} ddt
 */
export const addEventListeners = ddt => {
	if (typeof window !== 'undefined') {
		window.addEventListener('mousedown', ddt.onMouseDown);
		window.addEventListener('touchstart', ddt.onMouseDown);
		window.addEventListener('resize', ddt.onResize);
		if (ddt.props.repositionOnScroll) {
			window.addEventListener('scroll', ddt.onScroll);
		}
	}
};

/**
 * Remove the window event listeners that the DropDownToggled might have added.
 *
 * @export
 * @param {DropDownToggled} ddt
 */
export const removeEventListeners = ddt => {
	if (typeof window !== 'undefined') {
		window.removeEventListener('mousedown', ddt.onMouseDown);
		window.removeEventListener('touchstart', ddt.onMouseDown);
		window.removeEventListener('resize', ddt.onResize);
		window.removeEventListener('scroll', ddt.onScroll);
	}
};

/**
 * Throttles the execution of the given function to requestAnimationFrame.
 * This isn't a general solution since it depends on no args being passed, but
 * it's good enough for now
 *
 * @export
 * @param {Function} fn
 */
export const throttleToFrame = fn => {
	let requestingFrame = false;
	const wrapped = () => {
		fn();
		requestingFrame = false;
	};
	return () => {
		if (!requestingFrame) {
			window.requestAnimationFrame(wrapped);
			requestingFrame = true;
		}
	};
};
