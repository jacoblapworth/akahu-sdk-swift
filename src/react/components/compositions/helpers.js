import cn from 'classnames';
import { ns } from '../helpers/xuiClassNamespace';

const baseCompositionClass = `${ns}-composition`;

export const retainValues = {
	medium: 'medium',
	small: 'small',
};

export function buildGlobalCompositionClasses({
	isInfinite,
	hasGridGap,
	hasAutoSpaceAround,
	hasAutoColumnWidths,
}) {
	return cn(
		baseCompositionClass,
		!isInfinite && `${baseCompositionClass}-is-finite`,
		hasGridGap && `${baseCompositionClass}-has-grid-gap`,
		hasAutoSpaceAround && `${baseCompositionClass}-has-auto-space-around`,
		hasAutoColumnWidths && `${baseCompositionClass}-has-auto-widths`,
	);
}

export default baseCompositionClass;
