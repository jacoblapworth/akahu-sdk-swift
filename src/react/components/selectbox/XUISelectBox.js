import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import { nanoid } from 'nanoid';

import XUIDropdown from '../dropdown/XUIDropdown';
import XUIDropdownToggled from '../dropdown/XUIDropdownToggled';
import XUIButton from '../button/XUIButton';
import { sizes, widths } from './private/constants';
import XUIPicklist from '../picklist/XUIPicklist';
import qaHooks from './qaHooks';
import { ns } from '../helpers/xuiClassNamespace';
import XUIControlWrapper from '../controlwrapper/XUIControlWrapper';
import generateIds, { getAriaAttributes } from '../helpers/ariaHelpers';

import EditableTableCellContext from '../../contexts/EditableTableCellContext';

/**
 * If a qaHook is supplied in component props this helper provides a suffix for
 * sub-components
 *
 * @private
 * @param {String} propsQaHook The qaHook prop supplied to the component
 * @param {String} suffix      The suffix to add. This should be a constant from
 *                             the QA_HOOKS property of constants object.
 * @return {String | null}						 QA_HOOK + Suffix or null if qaHook is undefined
 */
function setQaHook(propsQaHook, suffix) {
  return propsQaHook ? `${propsQaHook}${suffix}` : null;
}

const selectBaseClass = `${ns}-select`;

export default class XUISelectBox extends Component {
  selectId = this.props.id || `xui-${nanoid(10)}`;

  trigger = React.createRef();

  ddt = React.createRef();

  wrapperIds = generateIds(this.selectId);

  isDropdownOpen = () => this.ddt.current?.isDropdownOpen();

  render() {
    return (
      <EditableTableCellContext.Consumer>
        {({ useCellStyling, cellAttributes }) => {
          const selectBox = this;
          const {
            buttonClassName,
            buttonContent,
            buttonVariant,
            caretTitle,
            children,
            closeAfterSelection,
            containerClassName,
            hasDefaultLayout,
            dropdownClassName,
            forceDesktop,
            fullWidth,
            hintMessage,
            inputGroupClassName,
            isDisabled,
            isFieldLayout,
            isInvalid,
            isLabelHidden,
            isOpen,
            isTextTruncated,
            label,
            labelClassName,
            matchTriggerWidth,
            onBlur,
            onDropdownHide,
            onFocus,
            onSelect,
            qaHook,
            restrictFocus,
            size,
            validationMessage,
          } = this.props;

          const buttonClassNames = cn(
            `${selectBaseClass}--button`,
            !buttonVariant && `${selectBaseClass}--button-no-variant`,
            buttonClassName,
          );
          const inputGroupClassNames = cn(
            `${selectBaseClass}`,
            hasDefaultLayout && `${selectBaseClass}-layout`,
            inputGroupClassName,
            !useCellStyling && isInvalid && `${selectBaseClass}-is-invalid`,
          );

          const containerClassNames = cn(
            `${selectBaseClass}wrapper`,
            useCellStyling && `${selectBaseClass}-cell`,
            !useCellStyling && isInvalid && `${selectBaseClass}wrapper-is-invalid`,
            containerClassName,
          );

          const ariaAttributes = cellAttributes || getAriaAttributes(this.wrapperIds, this.props);

          const trigger = (
            <XUIButton
              _caretClassName={`${selectBaseClass}--caret`}
              _useCellStyling={useCellStyling}
              {...ariaAttributes}
              className={buttonClassNames}
              fullWidth={fullWidth}
              hasCaret
              id={this.wrapperIds.control}
              isDisabled={isDisabled}
              onBlur={onBlur}
              onFocus={onFocus}
              qaHook={setQaHook(qaHook, qaHooks.button)}
              ref={selectBox.trigger}
              size={size}
              type="button"
              variant={buttonVariant}
            >
              <span
                className={cn(
                  `${selectBaseClass}--content`,
                  isTextTruncated && `${selectBaseClass}--content-truncated`,
                )}
              >
                {buttonContent}
              </span>
            </XUIButton>
          );

          const dropdown = (
            <XUIDropdown
              // These aria attributes currently go nowhere
              ariaAttributes={getAriaAttributes(this.wrapperIds.control, this.props)}
              className={dropdownClassName}
              id={selectBox.selectId}
              onSelect={onSelect}
              qaHook={setQaHook(qaHook, qaHooks.dropdown)}
              restrictFocus={restrictFocus}
            >
              <XUIPicklist>{children}</XUIPicklist>
            </XUIDropdown>
          );

          return (
            <div className={containerClassNames} data-automationid={qaHook}>
              <XUIControlWrapper
                isFieldLayout={isFieldLayout}
                qaHook={setQaHook(qaHook, '')} // label suffix is handled within controlwrapper
                wrapperIds={this.wrapperIds}
                {...{
                  isLabelHidden,
                  label,
                  isInvalid,
                  validationMessage,
                  hintMessage,
                  labelClassName,
                }}
              >
                <div
                  className={inputGroupClassNames}
                  data-automationid={setQaHook(qaHook, qaHooks.inputGroup)}
                >
                  {React.Children.count(children) === 0 ? (
                    trigger
                  ) : (
                    <XUIDropdownToggled
                      closeOnSelect={closeAfterSelection}
                      dropdown={dropdown}
                      forceDesktop={forceDesktop}
                      isBlock
                      isHidden={!isOpen}
                      matchTriggerWidth={matchTriggerWidth}
                      onClose={onDropdownHide}
                      qaHook={setQaHook(qaHook, qaHooks.dropdownToggled)}
                      ref={selectBox.ddt}
                      trigger={trigger}
                    />
                  )}
                </div>
              </XUIControlWrapper>
            </div>
          );
        }}
      </EditableTableCellContext.Consumer>
    );
  }
}

XUISelectBox.propTypes = {
  /** Additional classes to be applied to the button */
  buttonClassName: PropTypes.string,

  /** Display text to be rendered on the `XUISelectBox` button. */
  buttonContent: PropTypes.node.isRequired,

  /** The XUI button variant to use as a trigger for the select box */
  buttonVariant: PropTypes.string,

  /**
   * Title for the button caret
   * <br />
   * Recommended English value: *Toggle list*
   */
  caretTitle: PropTypes.string,

  children: PropTypes.node,

  /** When a selection is made, close the dropdown */
  closeAfterSelection: PropTypes.bool,

  /** Additional classes to be applied to the container */
  containerClassName: PropTypes.string,

  /** Additional classes to be applied to the dropDown */
  dropdownClassName: PropTypes.string,

  /** Force the desktop experience, even if the viewport is narrow enough for mobile */
  forceDesktop: PropTypes.bool,

  /**
   * Modifier for the width of the XUISelectBox: `always`, `small-down`, or `never`.
   */
  fullWidth: PropTypes.oneOf(widths),

  /** Use XUI provided layout classes */
  hasDefaultLayout: PropTypes.bool,

  /** Hint message to show under the input */
  hintMessage: PropTypes.node,

  /** ID to apply to the dropdown. Used primarily to associate a label with it's matched content.
   * If none is provided it's automatically generated. */
  id: PropTypes.string,

  /** Additional classes to be applied to the inputGroup */
  inputGroupClassName: PropTypes.string,

  /** Whether the button trigger and functionality are disabled */
  isDisabled: PropTypes.bool,

  /** Whether to use the field layout classes  */
  isFieldLayout: PropTypes.bool,

  /** Whether the current input value is invalid */
  isInvalid: PropTypes.bool,

  /** Input Label visibility */
  isLabelHidden: PropTypes.bool,

  /** Whether or not the list should be forced open */
  isOpen: PropTypes.bool,

  /** Optionally toggles the text truncation */
  isTextTruncated: PropTypes.bool,

  /** Input Label */
  label: PropTypes.node.isRequired,

  /** Additional classes to be applied to the label */
  labelClassName: PropTypes.string,

  /**
   * Setting this to `true` makes the dropdown as wide as the trigger (defaults to `true`). <br>
   * **Note:** *Setting this to true will override any size prop on Dropdown.* <br>
   * Setting this to `false` will allow the dropdown's width to be set independent of the trigger width. <br>
   * Setting this to `'min'` will set the dropdown's `min-width` to be the trigger width. <br/>
   * XUI design has also decided to keep a minimum width on the dropdown,
   * so dropdown may not match the width of narrow triggers (setting this to `'min'` will not override this).
   */
  matchTriggerWidth: PropTypes.oneOfType([PropTypes.bool, PropTypes.oneOf([true, false, 'min'])]),

  /** Optional callback to be executed when the trigger loses focus */
  onBlur: PropTypes.func,

  /** Optional callback to be executed when dropdown closes */
  onDropdownHide: PropTypes.func,

  /** Optional callback to be executed when the trigger gains focus */
  onFocus: PropTypes.func,

  /** Selection callback */
  onSelect: PropTypes.func,

  /** for adding automation ID to component as well as input and button sub-components */
  qaHook: PropTypes.string,

  /** Whether focus should be restricted to the dropdown while it's open. */
  restrictFocus: PropTypes.bool,

  /**
   * Modifier for the size of the select box: `medium`, `small`, or `xsmall`.
   *
   * If `XUISelectBoxOption` does not have a size set, it will inherit the size from `XUISelectBox`.
   */
  size: PropTypes.oneOf(sizes),

  /** Validation message to show under the input if `isInvalid` is true */
  validationMessage: PropTypes.node,
};

XUISelectBox.defaultProps = {
  closeAfterSelection: true,
  forceDesktop: false,
  fullWidth: 'always',
  hasDefaultLayout: true,
  isOpen: false,
  isTextTruncated: true,
  matchTriggerWidth: true,
  restrictFocus: true,
};
