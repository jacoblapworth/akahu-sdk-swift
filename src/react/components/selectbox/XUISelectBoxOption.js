import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import XUIPickitem from '../picklist/XUIPickitem';
import { ns } from '../helpers/xuiClassNamespace';

export default class XUISelectBoxOption extends PureComponent {
  render() {
    const {
      children,
      truncateText,
      onClick,
      onSelect,
      id,
      showCheckboxes,
      value,
      optionClassName,
      isDisabled,
      isSelected,
      isHighlighted,
      onBlur,
      onFocus,
      onKeyDown,
      href,
      ariaRole,
      onMouseOver,
      qaHook,
    } = this.props;
    const isText = typeof children[0] === 'string';
    const shouldTruncateChildren = isText && truncateText;
    const contents = shouldTruncateChildren ? (
      <span className={`${ns}-select--option-truncated`}>{children}</span>
    ) : (
      children
    );
    return (
      <XUIPickitem
        {...{
          onClick,
          onSelect,
          id,
          value,
          isSelected,
          isHighlighted,
          isDisabled,
          onBlur,
          onFocus,
          onKeyDown,
          href,
          ariaRole,
          onMouseOver,
          qaHook,
        }}
        className={optionClassName}
        isMultiselect={showCheckboxes}
      >
        {contents}
      </XUIPickitem>
    );
  }
}

XUISelectBoxOption.propTypes = {
  /** defaults to `option` for the aria role attribute, but can be defined for other uses. */
  ariaRole: PropTypes.string,

  children: PropTypes.node,

  /** link to be used in child, will render an a tag if used and button if not */
  href: PropTypes.string,

  id: PropTypes.string.isRequired,

  /** render the option disabled */
  isDisabled: PropTypes.bool,

  /** true if the item is highlighted */
  isHighlighted: PropTypes.bool,

  /** true if the item is selected */
  isSelected: PropTypes.bool,

  /** callback on blur of the pickitem */
  onBlur: PropTypes.func,

  /** callback when the pickitem is clicked. */
  onClick: PropTypes.func,

  /** callback on focus of the pickitem */
  onFocus: PropTypes.func,

  /** callback on keydown of the pickitem */
  onKeyDown: PropTypes.func,

  /** callback for mouseover event */
  onMouseOver: PropTypes.func,

  /** Callback to be executed after a selection is made */
  onSelect: PropTypes.func,

  /** Additional classes to be applied to the  option insides */
  optionClassName: PropTypes.string,

  /** The automation-id to add to the item */
  qaHook: PropTypes.string,

  /** Do the dropdown options show checkboxes */
  showCheckboxes: PropTypes.bool,

  /** Truncate `XUISelectBoxOption` children which are strings to one line */
  truncateText: PropTypes.bool,

  /** The value associated with this option */
  value: PropTypes.any.isRequired,
};

XUISelectBoxOption.defaultProps = {
  ariaRole: 'option',
  isDisabled: false,
  isHighlighted: false,
  isSelected: false,
  showCheckboxes: false,
  truncateText: false,
  /*
	 DO NOT REMOVE
	 This property is needed so that the stateful picklist will properly recognize this
	 component as a menu item.
	 */
  _isMenuItem: true, // eslint-disable-line
};
