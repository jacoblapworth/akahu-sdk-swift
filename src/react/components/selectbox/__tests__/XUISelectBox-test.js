import React from 'react';
import XUISelectBox from '../XUISelectBox';
import XUISelectBoxOption from '../XUISelectBoxOption';
import { then } from './helpers';
import Enzyme, { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import { v4 as uuidv4 } from 'uuid';
import { eventKeyValues } from '../../helpers/reactKeyHandler';

jest.mock('uuid');
uuidv4.mockImplementation(() => 'testSelectBoxId');

Enzyme.configure({ adapter: new Adapter() });

describe('<XUISelectBox />', function () {
  const options = ['Santa Cruz', 'Transition', 'Lapierre', 'Surly', 'Kona'];
  let select;
  let wasSelected = false;

  beforeEach(function () {
    select = mount(
      <XUISelectBox
        label="Test Select Box"
        value={options[0]}
        name="Test"
        buttonContent={options[0]}
        buttonClasses="blah"
        forceDesktop
        caretTitle="Toggle list"
      >
        {options.map((opt, idx) => {
          return (
            <XUISelectBoxOption
              id={opt}
              key={opt + idx}
              selected={opt === options[0]}
              onSelect={() => {
                wasSelected = true;
              }}
              value={opt}
              showCheckboxes={true}
              truncatedText={true}
            >
              {opt}
            </XUISelectBoxOption>
          );
        })}
      </XUISelectBox>,
    );
  });

  it('should place the defaultValue as the initial input value', function () {
    // Have to remove the title of the SVG used to show the down arrow icon
    const buttonValue = select.find('button').first().text().replace('Toggle list', '');
    expect(buttonValue).toEqual(options[0]);
  });

  it('should open the dropdown on click of the button when it has children', function () {
    select.find('button.blah').first().simulate('click');
    expect(select.instance().isDropdownOpen()).toBeTruthy();
  });

  it('should not open the dropdown on click of the button when it does not have children', function () {
    select = mount(
      <XUISelectBox
        value="Test"
        label="Does not have children"
        name="Test"
        buttonContent="test"
        buttonClasses="blah"
        isOpen={false}
        forceDesktop
        caretTitle="Toggle list"
      />,
    );

    select.find('button.blah').first().simulate('click');
    expect(select.instance().isDropdownOpen()).toBeFalsy();
  });

  it('should not open the dropdown on click if the control is disabled', function () {
    select = mount(
      <XUISelectBox
        label="test"
        buttonContent="test"
        buttonClasses="blah"
        isOpen={false}
        forceDesktop
        isDisabled
        caretTitle="Toggle list"
      >
        <XUISelectBoxOption id="1" value="A sample option" label="test">
          A sample option
        </XUISelectBoxOption>
      </XUISelectBox>,
    );

    select.find('button.blah').first().simulate('click');
    expect(select.instance().isDropdownOpen()).toBeFalsy();
  });

  it('should call onSelect on click of option', function () {
    then(() => select.find('.xui-pickitem').first().simulate('click')).then(() => {
      expect(wasSelected).toBeTruthy();
    });
  });

  it('should open the dropdown on arrowDown', function () {
    select.find('button.blah').first().simulate('keydown', {
      key: eventKeyValues.down,
      keyCode: 40,
      which: 40,
    });

    expect(select.instance().isDropdownOpen()).toBeTruthy();
  });

  it("should render the appropriate automation id's when a qaHook is provided", () => {
    const select = renderer.create(
      <XUISelectBox
        qaHook="test-selectbox"
        label="test"
        buttonContent="test"
        id="testThisSelect"
        forceDesktop
        caretTitle="Toggle list"
      >
        <XUISelectBoxOption
          id="1"
          value="A sample option"
          label="test"
          qaHook="test-selectboxoption"
        >
          A sample option
        </XUISelectBoxOption>
      </XUISelectBox>,
    );

    expect(select).toMatchSnapshot();
  });

  it('should render the trigger in a disabled state if `isDisabled` is set', () => {
    const select = renderer.create(
      <XUISelectBox
        label="test"
        buttonContent="test"
        forceDesktop
        isDisabled
        caretTitle="Toggle list"
      >
        <XUISelectBoxOption id="1" value="A sample option" label="test">
          A sample option
        </XUISelectBoxOption>
      </XUISelectBox>,
    );

    expect(select).toMatchSnapshot();
  });
});
