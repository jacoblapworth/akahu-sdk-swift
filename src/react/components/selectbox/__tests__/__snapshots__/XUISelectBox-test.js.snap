// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<XUISelectBox /> should render the appropriate automation id's when a qaHook is provided 1`] = `
<div
  className=""
  data-automationid="test-selectbox"
>
  <div
    className=""
  >
    <label
      className="xui-text-label xui-fieldlabel-layout"
      data-automationid="test-selectbox---label--label"
      htmlFor="testThisSelect-control"
    >
      test
    </label>
    <div
      className="xui-select-layout"
      data-automationid="test-selectbox---inputgroup"
    >
      <div
        className="xui-dropdown--toggledwrapper"
        data-automationid="test-selectbox---dropdowntoggled"
        data-ref="toggled-wrapper"
        role="presentation"
      >
        <button
          aria-activedescendant={null}
          aria-haspopup="listbox"
          className="xui-button xui-select--button xui-select--button-no-variant xui-button-standard xui-button-medium xui-button-fullwidth xui-button-has-icon"
          data-automationid="test-selectbox---button"
          disabled={false}
          id="testThisSelect-control"
          onClick={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            className="xui-select--content xui-select--content-truncated"
          >
            test
          </span>
          <div
            className="xui-button--caret xui-select--caret"
          >
            <svg
              className="xui-icon"
              focusable="false"
              height={5}
              viewBox="0 0 10 5"
              width={10}
            >
              <path
                d="M0 0l5 5 5-5z"
                role="presentation"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`<XUISelectBox /> should render the trigger in a disabled state if \`isDisabled\` is set 1`] = `
<div
  className=""
>
  <div
    className=""
  >
    <label
      className="xui-text-label xui-fieldlabel-layout"
      data-automationid={null}
      htmlFor="xui-testSelectBoxId-control"
    >
      test
    </label>
    <div
      className="xui-select-layout"
      data-automationid={null}
    >
      <div
        className="xui-dropdown--toggledwrapper"
        data-automationid={null}
        data-ref="toggled-wrapper"
        role="presentation"
      >
        <button
          aria-activedescendant={null}
          aria-haspopup="listbox"
          className="xui-button xui-select--button xui-select--button-no-variant xui-button-standard xui-button-medium xui-button-fullwidth xui-button-has-icon xui-button-is-disabled"
          data-automationid={null}
          disabled={true}
          id="xui-testSelectBoxId-control"
          onClick={[Function]}
          onKeyDown={null}
          onKeyUp={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="xui-select--content xui-select--content-truncated"
          >
            test
          </span>
          <div
            className="xui-button--caret xui-select--caret"
          >
            <svg
              className="xui-icon"
              focusable="false"
              height={5}
              viewBox="0 0 10 5"
              width={10}
            >
              <path
                d="M0 0l5 5 5-5z"
                role="presentation"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
  </div>
</div>
`;
